sleuth-camel-core/src/main/java/com.playtika.sleuth.camel/CreatedEventNotifier.java

```diff
@@ -0,0 +1,103 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.playtika.sleuth.camel;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.camel.Endpoint;
+import org.apache.camel.Exchange;
+import org.apache.camel.management.event.ExchangeCreatedEvent;
+import org.apache.camel.support.EventNotifierSupport;
+import org.springframework.cloud.sleuth.Span;
+import org.springframework.cloud.sleuth.SpanTextMap;
+import org.springframework.cloud.sleuth.Tracer;
+import org.springframework.cloud.sleuth.instrument.messaging.MessagingSpanTextMapExtractor;
+import org.springframework.cloud.sleuth.instrument.messaging.MessagingSpanTextMapInjector;
+import org.springframework.cloud.sleuth.util.SpanNameUtil;
+
+import java.util.EventObject;
+
+import static com.playtika.sleuth.camel.SleuthCamelConstants.FROM_CAMEL;
+
+@Slf4j
+@AllArgsConstructor
+public class CreatedEventNotifier extends EventNotifierSupport {
+
+    private static final String MESSAGE_COMPONENT = "camel";
+
+    private final Tracer tracer;
+    private final MessagingSpanTextMapExtractor spanExtractor;
+    private final MessagingSpanTextMapInjector spanInjector;
+
+    @Override
+    public void notify(EventObject event) {
+        log.trace("Caught event [{}] - processing...", event);
+        ExchangeCreatedEvent exchangeCreatedEvent = (ExchangeCreatedEvent) event;
+        Exchange exchange = exchangeCreatedEvent.getExchange();
+        Endpoint endpoint = exchange.getFromEndpoint();
+        CamelMessageTextMap carrier = new CamelMessageTextMap(exchange.getIn());
+
+        Span span = getSpan(carrier, endpoint);
+        span.tag(FROM_CAMEL, "true");
+
+        spanInjector.inject(span, carrier);
+    }
+
+    private Span getSpan(CamelMessageTextMap carrier, Endpoint endpoint) {
+        log.trace("Getting span for [{}] from endpoint [{}]...", carrier, endpoint);
+        Span spanFromMessage = buildSpan(carrier);
+        if (spanFromMessage != null) {
+            log.debug("Built span from message - {} from endpoint {}. Assuming it is remote one. Continuing it...",
+                    spanFromMessage, endpoint);
+            tracer.continueSpan(spanFromMessage);
+            spanFromMessage.logEvent(Span.SERVER_RECV);
+            return spanFromMessage;
+        } else {
+            String name = getSpanName(endpoint);
+            Span newSpan = tracer.createSpan(name);
+            newSpan.logEvent(Span.CLIENT_SEND);
+            log.debug("Message doesn't contain span data, new one was created - {} for endpoint.", newSpan, endpoint);
+            return newSpan;
+        }
+    }
+
+    private String getSpanName(Endpoint endpoint) {
+        return SpanNameUtil.shorten(MESSAGE_COMPONENT + "::" + endpoint.getEndpointKey());
+    }
+
+    private Span buildSpan(SpanTextMap carrier) {
+        try {
+            return this.spanExtractor.joinTrace(carrier);
```

Comment:
not necessarily here, but a later variant of this should be careful to only join a trace when it is an RPC (not a messaging span). If you accidentally call join on a multi-consumer span, it will be bad as one span will end up with junk in the UI from many endpoints... possibly not render at all

Reply:
iotw messaging span kinds like PRODUCER CONSUMER should be used when known

Reply:
well, I integrated camel to sleuth as messaging mechanism since it has pretty much similar behavior, so I create or extract all spans like from messaging, not RPC